cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manages-identities`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
+
+
+## Long term plan
+
+The long term plan is to make Operator managing CIDs the default, and deprecate the part of Agent that is currently used to manage CIDs.
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for pod creation. The control plane side of network policies and other features relying on CIDs will not carry out any required changes during this time, causing them to stop working.
+
+
+## Upgrade
+
+Safe upgrade, without downtime, requires simultaneous CID management compatibility.
+
+If we upgrade the Operator from a version that is not managing CIDs to a version that is managing CIDs while Agents are also managing CIDs, there should be no negative impact on the system. Everything should work the same.
+
+
+## Downgrade
+
+Downgrade without downtime is possible only if Agent daemonset is downgraded first -- updated to an earlier version where Agent manages CIDs. Otherwise, if Operator is downgraded first, there will be a downtime lasting at least the amount of time needed for Agent daemonset to be downgraded after the Operator downgrade.
+
+# Security
+
+* Agents will no longer need to have the RBAC permission for modifying CIDs.
+
+
+# Future enhancements
+
+
+## CID lazy creation
+
+Scalability and performance optimization
+
+Create CIDs only for labels used in network policies to greatly reduce the number of CIDs. Only pod labels used in the peer pod label selector of network policies will be relevant for CID creation.
+
+There are multiple levels of optimization and complexity this solution can go to. The proposal is to start with a basic implementation that will have effects like CID relevant labels, based on label keys from network policy selectors. Discussion of more advanced solutions is outside the scope of this document.
+
+
+## Operator acts on Pod instead of CEP changes
+
+Status: Included in the original proposal
+
+Performance optimization
+
+Operator doesn't watch for CEPs but watches for Pods instead. As soon as a pod is created (not started) Operator allocates a CID immediately. This way the new CID is getting created in parallel as the pod is getting scheduled.
+
+Downside: In case pods are unschedulable we will be allocating identities even though they might not be used at the end.
```

Comment:
This should be fairly rare. As long as the operator doesn't leak these but instead performs GC then this should be fine.

Reply:
Yes. I added that it's a very minor downside.